File: FSharp/Asterisk.FS/Swef/Astrobserve.fs
module Astrobserve

open System
open System.IO
open System.Text

open Sweph
open Sweph.Constants
open Factory
open Studio

let private init = 
    lazy
        swe_julday <- Sweph.swe_julday
        swe_jdut1_to_utc <- Sweph.swe_jdut1_to_utc
        true

do
    init.Force() |> ignore

let initTrigger = (do ()); true

let mutable private _opening_iflag : int    = (int Bag.SEFLG_SPEED) ||| (int Bag.SEFLG_MOSEPH)
let mutable private _positionType           = Bag.SEFLG_SIDEREAL
let mutable private _active_iflag  : int    = _opening_iflag ||| (int _positionType)   
let mutable private _houseSystem   : int    = Bag.EX_HOUSE_SYS_K
let mutable private _siderealMode  : int    = Bag.SE_SIDM_LAHIRI
let mutable private _ascendant     : int    = Bag.SE_ASC

let Open (ephemerisPath: string option) = 
    match ephemerisPath with
    | None -> ()
    | Some s -> 
        let ep = !~ s
        if Directory.Exists ep then
            Sweph.swe_set_ephe_path(ep) |> ignore
            _opening_iflag <- (_opening_iflag &&& (~~~ (int Bag.SEFLG_MOSEPH))) ||| (int Bag.SEFLG_SWIEPH)

let Close () =
    if Bag.SEFLG_SWIEPH = ((_opening_iflag |> Convert.ToInt64) &&& Bag.SEFLG_SWIEPH) then
        Sweph.swe_close()

type Config =
    { SiderealMode: int64; PositionType: int64; HouseSystem: int64; NodeType: int64 }
    static member Default() =
        let dv = { SiderealMode = int64 _siderealMode 
                   PositionType = _positionType 
                   HouseSystem  = int64 _houseSystem 
                   NodeType     = int64 nodeType }
        dv

let private _Setup (c: Config) =
    _siderealMode <- int c.SiderealMode
    Sweph.swe_set_sid_mode(_siderealMode, 0., 0.)

    _active_iflag <- _opening_iflag ||| (int c.PositionType) ||| (Bag.SEFLG_SPEED |> int)
    _houseSystem  <- int c.HouseSystem
    nodeType      <- int c.NodeType

    ()

let private _ObserveHouses (position: Position) (time: JulianDay) =
    let Jd jd as zx = time 
    let cusps = Array.zeroCreate<float> 13
    let ascmc = Array.zeroCreate<float> 10
    Sweph.swe_houses_ex(jd / 1.<dd>, _active_iflag, position.Latitude / 1.<deg>, position.Longitude / 1.<deg>, _houseSystem, cusps, ascmc) |> ignore
    (ascmc, cusps)

let private _ObserveAscendant (position: Position) (time: JulianDay) (asc: int) =
    let ascmc, cusps = _ObserveHouses position time
    ascmc.[asc] * 1.<deg>

let private _ObservePoint (id: Planet) (time: JulianDay) = 
    let xx = Array.zeroCreate<float> 6
    let serr = new StringBuilder(256)
    let Jd jd as zx = time
    Sweph.swe_calc_ut(jd / 1.<dd>, (id :> INumber<int>).Number(), _active_iflag, xx, serr) |> ignore
    // Sweph.swe_fixstar_ut(name, jd / 1.<dd>, _active_iflag |> Convert.ToInt64, xx, serr) |> Convert.ToInt32
    let pointPosition =
        match id with
        | Ke -> 
            let lon = (normalize (xx.[0] + 180.) 360.) * 1.<deg>
            let lat = xx.[1] * -1.<deg>
            { Longitude = lon; Latitude = lat } 
        | _ -> 
            let lon = xx.[0] * 1.<deg>
            let lat = xx.[1] * 1.<deg>
            { Longitude = lon; Latitude = lat } 
    { Id = id 
      Position = pointPosition 
      Flow = { Distance = xx.[2] * 1.<AU>; SpeedInLongitude = xx.[3] * 1.<deg/dd>; SpeedInLatitude = xx.[4] * 1.<deg/dd>; SpeedInDistance = xx.[5] * 1.<AU/dd> } }

let private _ObserveTransits (id: Planet) (position: Position) (time: JulianDay) (transitFlags: int32): JulianDay =
    let Jd jd as zx = time
    let tjd_ut = jd / 1.<dd>
    // ipl, starname <- id
    let epheflag = _active_iflag
    let rsmi = transitFlags
    let geopos = [| position.Longitude / 1.<deg>; position.Latitude / 1.<deg>; 0. |]
    let atpress = 0.
    let attemp = 0.
    let tret = ref 0.
    let serr = new StringBuilder(256)

    Sweph.swe_rise_trans(tjd_ut, (id :> INumber<int>).Number(), new StringBuilder(), epheflag, rsmi, geopos, atpress, attemp, tret, serr) |> ignore
    // Sweph.swe_rise_trans(tjd_ut, -1, name, epheflag, rsmi, geopos, atpress, attemp, tret, serr) |> ignore

    Jd <| !tret * 1.<dd>

type private pchan = Point chan
type private achan = float<deg> chan
type private tchan = JulianDay chan
type private schan = bool chan

/// channel batch ~ channels in a Go's select
type private AstroQuery = 
    | ForPlanet of Planet * JulianDay * pchan
    | ForAscendant of Position * JulianDay * int * achan
    | ForTransit of Planet * Position * JulianDay * (TransitFlag list) * tchan
    | ForSetup of Config * schan

let private astrobserve = 
    agent.Start(
        fun inbox ->
            let rec loop () = 
                async {
                    let! msg = inbox.Receive()

                    match msg with
                    | ForPlanet (id, t, chan) ->
                        do chan.Reply <| _ObservePoint id t
                    | ForAscendant (pos, t, asc, chan) ->
                        do chan.Reply <| _ObserveAscendant pos t asc
                    | ForTransit (pid, pos, t, trans, chan) ->
                        let transFlag = 
                            match trans with
                            | [] -> 
                                (TransitFlag.Rise :> INumber<int>).Number()
                            | _ ->
                                trans |> Seq.map (fun x -> (x :> INumber<int>).Number()) |> Seq.fold (fun x1 x2 -> x1 ||| x2) 0
                    
                        do chan.Reply <| _ObserveTransits pid pos t transFlag
                    | ForSetup (st, chan) ->
                        _Setup st
                        do chan.Reply <| true

                    return! loop ()
                }
            loop ())

let ObservePoint (id: Planet) (time: JulianDay) =
    let jd = time
    astrobserve.PostAndReply(fun c -> ForPlanet (id, jd, c))

let ObserveAscendant (pos: Position) (time: JulianDay) (asc: int) =
    let jd = time
    astrobserve.PostAndReply(fun chan -> ForAscendant (pos, jd, asc, chan))

let ObserveTransit pid position time transitFlags =
    astrobserve.PostAndReply(fun chan -> ForTransit (pid, position, time, transitFlags, chan))

let BasicSetup (c: Config) =
    astrobserve.PostAndReply(fun chan -> ForSetup (c, chan)) 

File: astrobserve/astrobserve.go
package astrobserve

import (
	"path/filepath"

	"github.com/dc0d/asterisk"
	"github.com/dc0d/gosweph"
)

func init() {
	go agent()
}

func IsDirect(speedInLongitude float64) bool {
	return speedInLongitude >= 0.
}

func IsNightly(asc, su float64) bool {
	sunLen := su
	ascLen := asc

	if sunLen < ascLen {
		sunLen = sunLen + 360.
	}

	if ascLen < sunLen && sunLen < ascLen+180. {
		return true
	} else {
		return false
	}
}

type ForPlanet struct {
	Planet    asterisk.Planet
	JulianDay asterisk.JulianDay
	Answer    chan asterisk.Point
}

type ForAscendant struct {
	Position  asterisk.Position
	JulianDay asterisk.JulianDay
	Asc       Angle
	Answer    chan float64
}

type ForTransit struct {
	Planet      asterisk.Planet
	Position    asterisk.Position
	JulianDay   asterisk.JulianDay
	TransitFlag asterisk.TransitFlag
	Answer      chan asterisk.JulianDay
}

type ForStop struct {
	Stop   bool
	Answer chan bool
}

var (
	PlanetChan    = make(chan ForPlanet, 1)
	AscendantChan = make(chan ForAscendant, 1)
	TransitChan   = make(chan ForTransit, 1)
	StopChan      = make(chan ForStop, 1)
)

func agent() {
	setup()
	defer dispose()
	for {
		select {
		case p := <-PlanetChan:
			pnt := observePoint(p.Planet, p.JulianDay)
			p.Answer <- pnt
		case a := <-AscendantChan:
			asc := observeAscendant(a.Position, a.JulianDay, a.Asc)
			a.Answer <- asc
		case t := <-TransitChan:
			trn := observeTransits(t.Planet, t.Position, t.JulianDay, int32(t.TransitFlag))
			t.Answer <- trn
		case st := <-StopChan:
			if st.Stop {
				st.Answer <- true
				return
			}
			//case <-time.After(time.Second): //TEMP:
			//	return
		}
	}
}

func dispose() {
	defer func() {
		if e := recover(); e != nil {
			// TODO:
		}
	}()

	gosweph.Swe_close()
}

func setup() {
	dir, err := filepath.Abs(filepath.Dir(asterisk.EphemerisPath))
	if err != nil {
		panic(err)
	}

	dir = filepath.Join(dir, asterisk.EphemerisPath)

	gosweph.Swe_set_sid_mode(int32(asterisk.SiderealMode), 0., 0.)
	gosweph.Swe_set_ephe_path(&dir)
}

func observeTransits(
	pln asterisk.Planet,
	position asterisk.Position,
	time asterisk.JulianDay,
	transitFlags int32) asterisk.JulianDay {
	jd := float64(time)
	tjd_ut := jd
	epheflag := asterisk.IFlag
	rsmi := transitFlags
	geopos := [3]float64{position.Lon, position.Lat, 0.}
	atpress := 0.
	attemp := 0.
	tret := 0.
	var serr string

	gosweph.Swe_rise_trans(tjd_ut, pln.Swe(), nil, epheflag, rsmi, &geopos, atpress, attemp, &tret, &serr)

	return asterisk.JulianDay(tret)
}

func observeAscendant(position asterisk.Position, time asterisk.JulianDay, asc Angle) float64 {
	_, ascmc := observeHouses(position, time)
	return ascmc[asc]
}

func observeHouses(position asterisk.Position, time asterisk.JulianDay) (cusps [13]float64, ascmc [10]float64) {
	jd := float64(time)
	_cusps := cusps
	_ascmc := ascmc
	gosweph.Swe_houses_ex(jd, asterisk.IFlag, position.Lat, position.Lon, int32(asterisk.HouseSystem), &_cusps, &_ascmc)

	cusps = _cusps
	ascmc = _ascmc
	return
}

func observePoint(pln asterisk.Planet, time asterisk.JulianDay) asterisk.Point {
	var xx [6]float64
	var serr string
	jd := float64(time)

	iflag := asterisk.IFlag
	gosweph.Swe_calc_ut(jd, pln.Swe(), iflag, &xx, &serr)

	var pointPosition asterisk.Position
	if pln == asterisk.Ke {
		lon := asterisk.PositiveMod(xx[0]+180., 360.)
		lat := xx[1] * -1.
		pointPosition = asterisk.Position{Lon: lon, Lat: lat}
	} else {
		lon := xx[0]
		lat := xx[1]
		pointPosition = asterisk.Position{Lon: lon, Lat: lat}
	}

	flow := asterisk.Flow{Distance: xx[2], SpeedInLongitude: xx[3], SpeedInLatitude: xx[4], SpeedInDistance: xx[5]}

	return asterisk.Point{ID: pln, Position: pointPosition, Flow: flow}
}
